model/riscv_insts_crypto.sail

```diff
@@ -0,0 +1,185 @@
+
+/*
+ * file: riscv_insts_crypto.sail
+ *
+ * This file contains instructions whos definitions do not change whether
+ * XLEN is 32 or 64.
+ *
+ * This file must be included in the model build whatever the value of XLEN.
+ *
+ */
+
+/*
+ * Scalar Cryptography Extension - Scalar        SHA256 instructions
+ * ----------------------------------------------------------------------
+ */
+
+/* AST clauses for the SHA256 instructions. */
+union   clause ast    =    SHA256SIG0 : (regidx,regidx)
+union   clause ast    =    SHA256SIG1 : (regidx,regidx)
+union   clause ast    =    SHA256SUM0 : (regidx,regidx)
+union   clause ast    =    SHA256SUM1 : (regidx,regidx)
+
+
+/* Encode/decode mappings for the SHA256 instructions. */
+mapping clause encdec =      SHA256SUM0   (rs1,rd) 
+    <-> 0b00 @ 0b01000 @ 0b00000 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+mapping clause encdec =      SHA256SUM1   (rs1,rd) 
+    <-> 0b00 @ 0b01000 @ 0b00001 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+mapping clause encdec =      SHA256SIG0   (rs1,rd) 
+    <-> 0b00 @ 0b01000 @ 0b00010 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+mapping clause encdec =      SHA256SIG1   (rs1,rd) 
+    <-> 0b00 @ 0b01000 @ 0b00011 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+
+/* Assembly string mappings for the SHA256 instructions. */
+mapping clause assembly = SHA256SIG0 (rs1,rd) <-> "sha256sig0" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+mapping clause assembly = SHA256SIG1 (rs1,rd) <-> "sha256sig1" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+mapping clause assembly = SHA256SUM0 (rs1,rd) <-> "sha256sum0" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+mapping clause assembly = SHA256SUM1 (rs1,rd) <-> "sha256sum1" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+val      crypto_sha256 : (sha256_op, regidx, regidx) -> Retired effect {escape,rreg,wreg}
+/* --- specification snippet begin --- */
```

Comment:
What are all these about?

Reply:
I'm sorry, I'm guessing at what you are asking about here. Did you mean what are these instructions for? Or why have I written the code this way? I hope the former...

These instructions are lightweight accelerators for the SHA2 hash function. They wrap up the single operand `sigma0/1` and `sum0/1` functions described in the SHA2 specs into a single instruction. They're explained more in section 3.3 of the [draft spec](https://github.com/riscv/riscv-crypto/releases/tag/v0.8.1-scalar).

Reply:
The specification snippet comments, not the valspecs or function definitions.

Reply:
If you need LaTeX, sail can already produce that as output (and we use that in our CHERI ISA specification). I don't know if there's a way to extract the raw sail snippets out, but matching on `^function` and the following `^}` in a simple shell/awk/sed/etc script should be easy enough.

Reply:
Well, there is an ongoing issue about RISC-V documentation all being moved into asciidoc, rather than LaTeX, as the base ISA, priv ISA, crypto and Bitmanip specs currently use. So, while I hope the SAIL tooling can be used for this, it's a bit of a TBD right now.

Reply:
```
awk '/^function\>/{f=$2; b=1} b{print>f".sail"} /^}/{b=0}' model/riscv_insts_crypto.sail
```
(requires GNU awk, so will need gawk if running on *BSD) will produce a .sail file containing the source for each function, provided you keep the whitespace sane. Or you can rewrite such a trivial script in your favourite language.

The main value add of saildoc is not so much extracting source code but its ability to insert LaTeX hyperlinks everywhere and to permit the use of markdown to document functions (like doxygen/javadoc/etc); for CHERI-RISC-V, all our instruction semantics documentation is done on the associated `execute` function clauses (so it's there in the model and stays in sync) and then pulled into our ISA spec as LaTeX. If you want that then you'll need to use LaTeX (or use something that lets you mix LaTeX with other markups) or write a separate saildoc backend for asciidoc.

Reply:
These still need deleting